File: 21/clojure/day2.clj
(defn parse-command [c]
  (let [[command x]
        (clojure.string/split c #" ")
        x (Integer. x)]
    (cond (= command "up") {:y (- x) :x 0}
          (= command "down") {:y x :x 0}
          (= command "forward") {:x x :y 0})))

(defn reducer [a b]
  {:x (+ (:x a) (:x b))
   :y (+ (:y a) (:y b))})


(defn reducer2 [a b]
  {:aim (+ (:aim a) (:y b))
   :x (+ (:x a) (:x b))
   :y (+ (:y a) (* (:aim a) (:x b)))
   })

(let [aseq  (map parse-command
                 (clojure.string/split (slurp *in*) #"\n"))
      part1 (reduce reducer {:x 0 :y 0} aseq)
      part2 (reduce reducer2 {:aim 0 :x 0 :y 0} aseq)]
  (println (* (:x part1) (:y part1)))
  (println (* (:x part2) (:y part2)))
  )

File: 21/rust/day2.rs
use std::io::{self, Read};

fn main() {
    let mut inp = String::new();
    io::stdin().read_to_string(&mut inp).unwrap();
    let mut inp = inp.split_whitespace();

    let mut v = Vec::new();

    while let Some(command) = inp.next() {
        let value: i32 = inp.next().unwrap().parse().unwrap();
        match command {
            "up" => v.push((0, -value)),
            "down" => v.push((0, value)),
            "forward" => v.push((value, 0)),
            _ => ()
        }
    }

    let (a, b) = v.iter().fold((0, 0), |(x, y), (a, b)| (x+a, y+b));
    println!("Part1: {}", a*b);

    let (_, x, y) = v.iter().fold((0, 0, 0), 
        |(aim, x1, y1), (x2, y2)| (aim+y2, x1+x2, y1+aim*x2)
    );
    println!("Part2: {}", x*y);
}